src/modules/ekf2/EKF/aid_sources/optical_flow/optical_flow_control.cpp

```diff
@@ -90,20 +114,30 @@ void Ekf::controlOpticalFlowFusion(const imuSample &imu_delayed)
 				&& is_within_max_sensor_dist;
 
 		const bool starting_conditions_passing = continuing_conditions_passing
+				&& is_quality_good
+				&& is_magnitude_good
+				&& is_tilt_good
 				&& isTimedOut(_aid_src_optical_flow.time_last_fuse, (uint64_t)2e6); // Prevent rapid switching
 
 		// If the height is relative to the ground, terrain height cannot be observed.
-		_hagl_sensor_status.flags.flow = _control_status.flags.opt_flow && !(_height_sensor_ref == HeightSensor::RANGE);
+		_control_status.flags.opt_flow_terrain = _control_status.flags.opt_flow && !(_height_sensor_ref == HeightSensor::RANGE);
 
 		if (_control_status.flags.opt_flow) {
 			if (continuing_conditions_passing) {
-				fuseOptFlow(_hagl_sensor_status.flags.flow);
+
+				if (is_quality_good && is_magnitude_good && is_tilt_good) {
+					fuseOptFlow(H, _control_status.flags.opt_flow_terrain);
+				}
 
 				// handle the case when we have optical flow, are reliant on it, but have not been using it for an extended period
 				if (isTimedOut(_aid_src_optical_flow.time_last_fuse, _params.no_aid_timeout_max)) {
-					if (is_flow_required) {
+					if (is_flow_required && is_quality_good) {
```

Comment:
So, if `!is_magnitude_good || !is_tilt_good`, the flow data is never fused above but we reset every `no_aid_timeout_max`?

Reply:
Yes, for context I'm trying to avoid the situation we have with the current logic where we can be stuck with `_control_status.flags.opt_flow = true`, but with no actual flow fusion happening, and not hitting timeout logic. Then you can imagine the various implications in other areas of the EKF with "flow active", but silently doing nothing.

The specifics of what to do in different scenarios is what we'll probably need to review and adjust. Things like if you're briefly exceeding the tilt limit or quality min how long should flow fusion still remain "active" before timeout.